### PR TITLE
 BUG/ENH: Create npy format 3.0 to support extended unicode characters in dtypes

### LINKED ISSUE #7391: Can't save .npy file when array's field contains Chinese
It seem's numpy can't save .npy file when array's fields contains Chinese

``` python
import numpy as np
f=np.array([(1,3,3),
               (1, 2, [1,2]), 
               (1, 3, {1:2, 3:3, 4:34}), 
               (1, 2, 'ss')], 
              dtype=([('int', int), ('整形', int), ('对象', np.object_)]))
f['整形']    # It work's well with field contains Chinese
#->  array([  3,   2, 324,   2]) 
np.save('f.npy',f)  # But can't save it
```

I got following An error:

```
Traceback (most recent call last):
  File "<ipython-input-5-d9dd063afc64>", line 1, in <module>
    np.save('f.npy',f)  # But can't save it
  File ".../numpy/lib/npyio.py", line 491, in save
    pickle_kwargs=pickle_kwargs)
  File ".../numpy/lib/format.py", line 555, in write_array
    version)
  File ".../numpy/lib/format.py", line 314, in _write_array_header
    header = asbytes(_filter_header(header))
  File ".../numpy/compat/py3k.py", line 31, in asbytes
    return str(s).encode('latin1')
UnicodeEncodeError: 'latin-1' codec can't encode characters in position 29-30: ordinal not in range(256)
```

Then I try

``` python
f.dtype.names=('a','b','c')
np.save('f.npy',f)
```

Everything turns OK

Numpy Version = '1.10.4'
Python 3.4.4
win7

### PR DESCRIPTION
**Pull request by eric-wieser**
This version encodes the dtype as utf8 instead of latin1.

Unfortunately we need to create a new version to make this change, because we did not limit ourselves to ASCII in versions 1 and 2.

Fixes gh-7391

---

Two commits - a refactor to make adding new formats easy, and then the new format. Can split if it's helpful

### CODE REVIEW

File: numpy/lib/format.py [hunk 1]
```diff
@@ -162,7 +169,7 @@
 import warnings
 from numpy.lib.utils import safe_eval
 from numpy.compat import (
-    asbytes, asstr, isfileobj, long, os_fspath, pickle
+    isfileobj, long, os_fspath, pickle
```
Comment (eric-wieser) [inline numpy/lib/format.py]: I think I'd argue that every occurrence of `asbytes` and `asstr` is probably a bug, especially once we drop py2

File: numpy/lib/tests/test_format.py [hunk 2]
```diff
@@ -882,3 +883,27 @@ def test_empty_npz():
...
+def test_unicode_field_names():
+    # gh-7391
+    arr = np.array([
+        (1, 3),
+        (1, 2),
+        (1, 3),
+        (1, 2)
+    ], dtype=[
+        ('int', int),
+        (u'\N{CJK UNIFIED IDEOGRAPH-6574}\N{CJK UNIFIED IDEOGRAPH-5F62}', int)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_format.py]: Didn't want to open the mess of putting weird characters in our source code, but wanted to reproduce the issue faithfully.

File: numpy/lib/format.py [hunk 3]
```diff
@@ -603,12 +636,7 @@ def write_array(fp, array, version=None, allow_pickle=True, pickle_kwargs=None):
 
     """
     _check_version(version)
-    used_ver = _write_array_header(fp, header_data_from_array_1_0(array),
-                                   version)
-    # this warning can be removed when 1.9 has aged enough
-    if version != (2, 0) and used_ver == (2, 0):
-        warnings.warn("Stored array in format 2.0. It can only be"
-                      "read by NumPy >= 1.9", UserWarning, stacklevel=2)
```
Comment (eric-wieser) [inline numpy/lib/format.py]: These warnings moved into `_wrap_header_guess_version`

File: numpy/lib/format.py [hunk 4]
```diff
@@ -519,16 +553,15 @@ def _read_array_header(fp, version):
...
-    else:
+    hinfo = _header_size_info.get(version)
+    if hinfo is None:
         raise ValueError("Invalid version {!r}".format(version))
+    hlength_type, encoding = hinfo
 
     hlength_str = _read_bytes(fp, struct.calcsize(hlength_type), "array header length")
     header_length = struct.unpack(hlength_type, hlength_str)[0]
     header = _read_bytes(fp, header_length, "array header")
+    header = header.decode(encoding)
```
Comment (eric-wieser) [inline numpy/lib/format.py]: Previously this decode happened within `_filter_header`

**Comment by mattip** [comment]
LGTM. Needs a merge conflict resolution. 

It seems we get many of these where the release note gets conflicted, I wonder if we should move to a system where the release note is built up of many smaller files. I am sure I saw a project like that somewhere but googling "tool to build release notes" is not trivial

**Comment by mattip** [comment]
Does adding a new format number need to hit the mailing list?

**Comment by eric-wieser** [comment]
Town crier has  been suggested in the past, I think it's what matplotlib uses. I think it would be a great thing to move to. I think @rgommers might have been the one to bring it up.

**Comment by eric-wieser** [comment]
My opinion is the number does not need to hit the list, the only people who will ever see the new number are the ones who previously could not save their data at all.

**Comment by tylerjereddy** [comment]
SciPy just uses a wiki page for drafting the notes before release time to avoid merge conflicts. (Eventual) more elegant / automated solutions would be nice, especially if human intervention on the release note curation process can be reduced once the hiccups are sorted out.

**Comment by mattip** [comment]
Putting this in. Should we backport the new format?

**Comment by charris** [comment]
> Should we backport the new format?

Probably worth a try, it would be nice to have the extra compatibility.

File: numpy/lib/format.py (hunk 1 above)
Comment (charris) [reply numpy/lib/format.py]: Doesn't backport.

Comment (charris) [reply numpy/lib/format.py]: There is a reason we are putting off those cleanups.

**Comment by eric-wieser** [comment]
> EDIT: I don't trust it to work with Python 2.7, so dropping backport.

I tried backporting anyway - what's your concern on python 2?

**Comment by charris** [comment]
The removal of `asbytes` and `asstr`.

**Comment by eric-wieser** [comment]
You mean the fact they are no longer exported (shouldn't matter, they should be imported from `np.compat` by users), or the fact they're no longer used?

Assuming your concern is the latter, those functions have essentially just been inlined in this PR.

**Comment by charris** [comment]
> Assuming your concern is the latter, those functions have essentially just been inlined in this PR.

That was my concern.

**Comment by charris** [comment]
@eric-wieser What is the status of your review of this?

**Pull request merged**